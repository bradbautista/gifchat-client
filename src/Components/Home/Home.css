body {
    background-color: lightblue;
}
   
header {
  background-color: seashell;
  margin: 0;
  display: flex;
  justify-content: center;
  border-bottom: 1px solid purple;
  padding-top: 1rem;
  padding-bottom: 1rem;
}

h1 {
  font-size: 6.5rem;
  margin: 0;
}

main {
  background-color: seashell;
  opacity: 0.99;
}

section {
  display: flex;
  flex-direction: column;
  padding: 4rem;
}

.room-link {
  font-size: 2.5rem;
  font-weight: 900;
  line-height: 2.2rem;
  margin-top: 1rem;
  margin-bottom: 1rem;
}

.hidden {
  display: none;
}

.home-input-flex-wrapper {
  display: flex;
  flex-direction: column;
}

.destination-field {
  height: 1rem;
  padding: 2.5rem;
  font-size: 2rem;
  width: 100%;
  margin-bottom: 2rem;
}

.go-button {
  background: rgb(130, 224, 255); 
  border: none;
  width: 100%;
  height: 7.7rem;
  padding: 0;
  margin: 0;
}

h2 {
  margin: 0;
  font-size: 3.8rem;
  line-height: 3.2rem;
  margin-bottom: 1rem;
  margin-top: 1rem;
  font-weight: 900;
}

ul {
  margin: 0;
  padding: 0;
}

strong {
  font-weight: 900;
}

li {
  margin: 0;
  margin-bottom: 1.5rem;
  padding: 0;
  /* padding-left: 0.5rem; */
  font-size: 2rem;
}

footer {
  background-color: gainsboro;
  display: flex;
  justify-content: space-evenly;
  padding: 1rem;
  font-size: 2rem;
}

  
/* TABLET BREAKPOINT */

@media screen and (min-width: 720px) {
  
  body {
    background-color: darkblue;
  }

  section {
      padding: 8rem;
  }

  .room-button {
      padding: 4rem;
      font-size: 3rem;
      margin-bottom: 1rem;
  }
}

/* DESKTOP BREAKPOINT */

@media screen and (min-width: 1024px) {

  body {
    background-color: seashell;
  }

  header {
      justify-content: flex-start;
      padding-left: 8rem;
      padding-bottom: 2rem;
      padding-top: 2rem;
  }

  section {
      padding-top: 5rem;
      width: 80rem;
      padding-bottom: 4rem;
      padding-left: 8rem;
  }

  .room-link {
    font-size: 3.4rem;
    font-weight: 900;
    line-height: 3rem;
  }

  h1 {
    font-size: 12rem;
    font-weight: 900;
    color: cadetblue;
    
  }

  h2 {
      margin-top: 3rem;
      margin-bottom: 2rem;
  }

  .home-input-flex-wrapper {
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
  }

  .destination-field {
    height: 1rem;
    padding: 2.5rem;
    font-size: 2rem;
    width: 75%;
    margin-bottom: 0;
  }

  .go-button {
    background: rgb(130, 224, 255); 
    width: 20%;
    height: 5.2rem;
  }

  footer {
      justify-content: flex-start;
  }

  footer a {
      margin-left: 8rem;
  }


}