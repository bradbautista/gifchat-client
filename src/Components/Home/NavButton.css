.card-container {
    height: 26rem;
    margin-bottom: 5rem;
    position: relative;
}

.card {
    display: flex;
    flex-direction: column;
    justify-content: center;
    position: absolute;
    padding: 3rem;
    width: 100%;
    outline: 2px solid var(--secondary);
    z-index: 5;
    cursor: pointer;
}
  
.front {
    background-color: var(--quaternary);
    z-index: 0;
    bottom: 0px;
    top: 0px;
    left: 0px;
    right: 0px;
    font-size: 3rem;
    font-weight: 900;
    color: black;
}
  
.back {
    background-color: var(--quinary);
    z-index: -6;
    top: -10px;
    left: -10px;
    bottom: 10px;
    right: 10px;
    font-size: 2rem;
    transition: 0.5s;
    letter-spacing: -0.2px;
}
  
.swap {
    transition: all 0.5s;
    z-index: 6;
    top: 10px;
    left: 10px;
    right: -10px;
    bottom: -10px;
}

/* TABLET BREAKPOINT */
  
@media screen and (min-width: 720px) {

    .front {
        font-size: 4.5rem;
        padding-left: 7rem;
    }

    .back {
        font-size: 2.5rem;
    }
    
}

/* iPad pro BREAKPOINT */

@media screen and (min-width: 1024px) {

    .card-container {
        height: 30rem;
    }

    .front {
        font-size: 6rem;
        padding-left: 7rem;
    }

    .back {
        font-size: 3.5rem;
    }

}


/* DESKTOP BREAKPOINT */
  
@media screen and (min-width: 1200px) {

    .card-container {
        height: 17rem;
        margin-bottom: 5rem;
    }

    .front {
        font-size: 3rem;
    }

    .back {
        font-size: 2rem;
    }
    
}