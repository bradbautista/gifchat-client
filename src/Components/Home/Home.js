import React, { Component } from 'react'
import { Link } from 'react-router-dom'
import './Home.css'
import config from '../../config'
import NavButton from './NavButton'

export default class Home extends Component {

  // It'll arise naturally but think about what state we might need to track on this page if any; may just come in the form of context provision

  constructor(props) {
    super(props);
    this.state = {
      destination: '',
      createdRoom: '',
      lonelyRandos: false,
    };

  }

  getRoom = () => {

    console.log('getRoom fired')

    // >> user clicks "Get a room"

    // Send a POST request to the server (/ or /rooms/ or /requestRoom)
    // server should create a room at a URL and respond with that URL
    // 
    
    const endpoint = config.GIFCHAT_API_ENDPOINT
    const url = `${endpoint}rooms/`
    const options = {
      method: 'POST',
      body: '',
      headers: {
        'content-type': 'application/json',
      }
    }

    // console.log(url)
    // console.log(endpoint)
    // console.log(query)
    // console.log(this.state)

    return fetch(url, options)
        .then((res) => {
          return res.json();
        })
        .then((responseJson) => {

          console.log(responseJson)
          console.log(responseJson[0])

          this.setState({
            createdRoom: responseJson[0]
          })
          
        })
        .catch(error => { console.error(error) })

  }

  satisfyReact = () => {
  }

  handleChange = (e) => {
    
    this.setState({destination: e.target.value});
  
  }

  render() {

    return (
        <>
        <header>
          <h1>GifChat</h1>
        </header>
        
        <main>
            <section>

                {/* Ultimately we want a function invocation on button click,
                and a different one for each button. So let's walk through
                what everything needs to do:
                
                When "Get a room" is clicked, a link is generated by the server,
                it will requestRoom() -- send a POST request to the server to generate a new room

                Copy to clipboard?
                
                
                */}

              {/* GET ROOM */}

              <NavButton
                onClick={this.getRoom}
                frontCardContent="Get a room" 
                backCardContent={

                  // Feels hacky, but: Combining strings ('Room created at...', etc.) and Link object caused a lot of [Object object], so instead we need to do it in HTML. Because we need to do it in HTML, we need to return only one element instead of three (span, link, span).

                  <>
                    <span>Room created at </span>
                      <Link 
                        to={`/rooms/${this.state.createdRoom}`}
                        className='room-link'
                      >
                        {this.state.createdRoom}.
                        
                      </Link>
                    <span> Link copied to clipboard!</span>
                  </>

                }
              />

              {/* GO TO ROOM */}

              <NavButton

                // This is dumb but it works. So, NavButton is passing this.props.onClick and this.toggleSwap onClick, so React gets mad if there's not an onClick function assigned. But because of the fact that we're passing two functions and that requires a semicolon to work, JSX gets mad if we try to conditionally pass the onClick props based on this.props.frontCardContent. Thus, we pass this onClick an empty function and the program is happy.
                onClick={() => this.satisfyReact}
                frontCardContent='Go to room'
                backCardContent={
                  <div className='home-input-flex-wrapper'>
                    <input 
                      className='destination-field' 
                      value={this.state.destination} 
                      onChange={this.handleChange} 
                      type="text" placeholder="Enter URL"/>
                    <button className='go-button'
                      onClick={() => {console.log('Yo')}}
                    >Go</button>
                  </div>
                }

              />

              {/* GET A RANDO */}
                <NavButton 
                  onClick={() => {this.props.history.push('/randos/')}}
                  frontCardContent='Get a rando'
                  backCardContent=''
                />
        

              {/* HOW TO // Should this be a component? */}
              <h2>Here's how this works</h2>
              <ul>
                  <li><strong>No logins.</strong> GifChat does not want your email address.</li>
                  <li><strong>No names.</strong> GifChat does not care who you are.</li>
                  <li><strong>Two to a room.</strong></li>
                  <li><strong>GIFs only.</strong></li>
                  <li><strong>Conversations last while they're active.</strong> Your room URL is your link to that conversation. If conversations go inactive for seven days, they disappear and the room is closed.</li>
              </ul>
           
            </section>


                


        </main>

        <footer>
          {/* We'll also need to import FontAwesome icons or just use text links */}
          <a href="meongithub">FA GH</a> <a href="meonlinkedin">FA LI</a>
        </footer>

        </>
    );
  }  
}
