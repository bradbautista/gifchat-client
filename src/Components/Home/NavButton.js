import React, { Component } from 'react'
import './NavButton.css'

export default class NavButton extends Component {

  constructor(props) {
    super(props);
    this.state = {
      swapped: false,
    };

  }

  toggleSwap = () => {
    const currentState = this.state.swapped;
    this.setState({ swapped: !currentState})
  }

  render() {

    return (        

        <div className="card-container">
            <div 
            onClick={() => {this.props.onClick(); this.toggleSwap()}}
            // onClick={(this.props.frontCardContent == 'Get a rando') ? this.props.onClick : this.toggleSwap} 
            className="card front"
            >
                {this.props.frontCardContent}
            </div>

            <div 
            // onClick={this.toggleSwap} 
            className={'card' + ' ' + 'back' + ' ' + ((this.state.swapped) ? 'swap' : null)}>
                {this.props.backCardContent}
            </div>
        </div>



            
    );
  }  
}
