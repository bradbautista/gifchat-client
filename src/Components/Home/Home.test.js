import React from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter } from 'react-router-dom';
import { Link } from 'react-router-dom';
import Home from './Home';
import renderer from 'react-test-renderer';
import NavButton from './NavButton';

// Router wraps everything in index.js,
// but we need to wrap it here else the test complains
// about not having one

it('Renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(<BrowserRouter><Home /></BrowserRouter>, div)
  ReactDOM.unmountComponentAtNode(div);
});


it('Renders the UI as expected', () => {
    const someFunc = () => {};
    const someStr = 'a-big-red-dog-named-Clifford';
    const tree = renderer
        .create(
            <BrowserRouter>
                <header>
                <h1>GifChat</h1>
                </header>

                <main>
                <section>
                    {/* GET ROOM */}

                    <NavButton
                    onClick={someFunc}
                    frontCardContent="Get a room"
                    backCardContent={
                        <>
                        <span>Room created at </span>
                        <Link
                            to={`/rooms/${someStr}`}
                            className="room-link"
                            target="_blank"
                            rel="noopener noreferrer"
                        >
                            {someStr}.
                        </Link>
                        </>
                    }
                    />

                    {/* GO TO ROOM */}

                    <NavButton
                    // This is dumb but it works. So, NavButton is passing this.props.onClick and this.toggleSwap onClick, so React gets mad if there's not an onClick function assigned. But because of the fact that we're passing two functions and that requires a semicolon to work, JSX gets mad if we try to conditionally pass the onClick props based on this.props.frontCardContent. Thus, we pass this onClick an empty function and the program is happy.
                    onClick={someFunc}
                    frontCardContent="Go to room"
                    backCardContent={
                        <>
                        <span className="input-prompt">
                            Enter room name, i.e. a-big-red-dog-named-Clifford
                        </span>
                        <div className="home-input-flex-wrapper">
                            <input
                            className="destination-field"
                            value={someStr}
                            onChange={someFunc}
                            type="text"
                            placeholder="a-big-red-dog-named-Clifford"
                            />
                            <button
                            className="go-button"
                            onClick={() => {
                                this.openInNewTab(`rooms/${someStr}`);
                            }}
                            >
                            Go
                            </button>
                        </div>
                        </>
                    }
                    />

                    {/* GET A RANDO */}
                    <NavButton
                    onClick={() => {
                        someFunc;
                    }}
                    frontCardContent="Get a rando"
                    backCardContent="Searching for randos. If none are found, you will be put in a room to wait until one comes along."
                    />

                    {/* HOW TO */}
                    <section className="explainer">
                    <h2>Rules of GifChat</h2>
                    <ul>
                        <li>
                        <strong>No logins.</strong> GifChat does not want your email
                        address.
                        </li>
                        <li>
                        <strong>No names.</strong> GifChat does not care who you are.
                        </li>
                        <li>
                        <strong>Two to a room.</strong>
                        </li>
                        <li>
                        <strong>GIFs only.</strong>
                        </li>
                        <li>
                        <strong>Conversations last while they're active.</strong> Your room URL is your link to that conversation. If conversations go inactive for seven days, they disappear and the room is closed.
                        </li>
                        <li>
                        <strong>Why?</strong> <a href="https://github.com/bradbautista/gifchat-client#readme" target="_blank" rel="noopener noreferrer"><strong>Because.</strong></a>
                        </li>
                    </ul>
                    </section>

                </section>
                </main>

                <footer>
                {' '}
                <a href="https://github.com/bradbautista/" target="_blank" rel="noopener noreferrer">
                    <i className="fab fa-github"></i>
                </a>
                <a href="https://www.linkedin.com/in/bradford-bautista/" target="_blank" rel="noopener noreferrer">
                    {' '}
                    <i className="fab fa-linkedin-in"></i>
                </a>
                </footer>
            </BrowserRouter>
            )
        .toJSON();
    expect(tree).toMatchSnapshot();
        
})