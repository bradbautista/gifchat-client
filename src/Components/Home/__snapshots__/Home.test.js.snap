// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders the UI as expected 1`] = `
Array [
  <header>
    <h1>
      GifChat
    </h1>
  </header>,
  <main>
    <section>
      <div
        className="card-container"
      >
        <div
          className="card front"
          onClick={[Function]}
        >
          Get a room
        </div>
        <div
          className="card back null"
        >
          <span>
            Room created at 
          </span>
          <a
            className="room-link"
            href="/rooms/a-big-red-dog-named-Clifford"
            onClick={[Function]}
            rel="noopener noreferrer"
            target="_blank"
          >
            a-big-red-dog-named-Clifford
            .
          </a>
        </div>
      </div>
      <div
        className="card-container"
      >
        <div
          className="card front"
          onClick={[Function]}
        >
          Go to room
        </div>
        <div
          className="card back null"
        >
          <span
            className="input-prompt"
          >
            Enter room name, i.e. a-big-red-dog-named-Clifford
          </span>
          <div
            className="home-input-flex-wrapper"
          >
            <input
              className="destination-field"
              onChange={[Function]}
              placeholder="a-big-red-dog-named-Clifford"
              type="text"
              value="a-big-red-dog-named-Clifford"
            />
            <button
              className="go-button"
              onClick={[Function]}
            >
              Go
            </button>
          </div>
        </div>
      </div>
      <div
        className="card-container"
      >
        <div
          className="card front"
          onClick={[Function]}
        >
          Get a rando
        </div>
        <div
          className="card back null"
        >
          Searching for randos. If none are found, you will be put in a room to wait until one comes along.
        </div>
      </div>
      <section
        className="explainer"
      >
        <h2>
          Rules of GifChat
        </h2>
        <ul>
          <li>
            <strong>
              No logins.
            </strong>
             GifChat does not want your email address.
          </li>
          <li>
            <strong>
              No names.
            </strong>
             GifChat does not care who you are.
          </li>
          <li>
            <strong>
              Two to a room.
            </strong>
          </li>
          <li>
            <strong>
              GIFs only.
            </strong>
          </li>
          <li>
            <strong>
              Conversations last while they're active.
            </strong>
             Your room URL is your link to that conversation. If conversations go inactive for seven days, they disappear and the room is closed.
          </li>
          <li>
            <strong>
              Why?
            </strong>
             
            <a
              href="https://github.com/bradbautista/gifchat-client#readme"
              rel="noopener noreferrer"
              target="_blank"
            >
              <strong>
                Because.
              </strong>
            </a>
          </li>
        </ul>
      </section>
    </section>
  </main>,
  <footer>
     
    <a
      href="https://github.com/bradbautista/"
      rel="noopener noreferrer"
      target="_blank"
    >
      <i
        className="fab fa-github"
      />
    </a>
    <a
      href="https://www.linkedin.com/in/bradford-bautista/"
      rel="noopener noreferrer"
      target="_blank"
    >
       
      <i
        className="fab fa-linkedin-in"
      />
    </a>
  </footer>,
]
`;
